module/state_synchronization/indexer/in_memory_indexer.go

```diff
@@ -0,0 +1,221 @@
+	Registers    *unsynchronized.Registers
+	Events       *unsynchronized.Events
+	Collections  *unsynchronized.Collections
+	Transactions *unsynchronized.Transactions
+	Results      *unsynchronized.LightTransactionResults
+}
+
+// InMemoryIndexer handles indexing of block execution data in memory.
+// It stores data in unsynchronized in-memory caches that are designed
+// to be populated once before being read.
+type InMemoryIndexer struct {
+	log          zerolog.Logger
+	metrics      module.ExecutionStateIndexerMetrics
+	registers    *unsynchronized.Registers
+	events       *unsynchronized.Events
+	collections  *unsynchronized.Collections
+	transactions *unsynchronized.Transactions
+	results      *unsynchronized.LightTransactionResults
+}
+
+// NewInMemoryIndexer creates a new indexer that uses in-memory storage implementations.
+// This is designed for processing unsealed blocks in the optimistic syncing pipeline.
+// The caches are created externally and passed to the indexer, as they will also be used
+// by the persister to save data permanently when a block is sealed.
+func NewInMemoryIndexer(config InMemoryIndexerConfig) *InMemoryIndexer {
+	indexer := &InMemoryIndexer{
+		log:          config.Log.With().Str("component", "in_memory_indexer").Logger(),
+		metrics:      config.Metrics,
+		registers:    config.Registers,
+		events:       config.Events,
+		collections:  config.Collections,
+		transactions: config.Transactions,
+		results:      config.Results,
+	}
+
+	indexer.metrics.InitializeLatestHeight(indexer.registers.LatestHeight())
+
+	indexer.log.Info().
+		Uint64("first_height", indexer.registers.FirstHeight()).
+		Uint64("latest_height", indexer.registers.LatestHeight()).
+		Msg("indexer initialized")
+
+	return indexer
+}
+
+// RegisterValue retrieves register values by the register IDs at the provided block height.
+// If a register is not found it will return a nil value and not an error.
+// Expected errors:
+// - storage.ErrHeightNotIndexed if the given height was not indexed yet or lower than the first indexed height.
+func (i *InMemoryIndexer) RegisterValue(ID flow.RegisterID, height uint64) (flow.RegisterValue, error) {
+	value, err := i.registers.Get(ID, height)
+	if err != nil {
+		// only return an error if the error doesn't match the not found error, since we have
+		// to gracefully handle not found values and instead assign nil, that is because the script executor
+		// expects that behavior
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	return value, nil
+}
+
+// IndexBlockData indexes all execution block data.
+// Unlike the original implementation, this doesn't use batches as the in-memory
+// caches don't support batched writes.
+func (i *InMemoryIndexer) IndexBlockData(data *execution_data.BlockExecutionDataEntity) error {
+	log := i.log.With().Hex("block_id", logging.ID(data.BlockID)).Logger()
+	log.Debug().Msgf("indexing block data")
+
```

Comment:
I think the indexer should take the execution result as a constructor argument, then verify here that `data.BlockID` matches the blockID from the result